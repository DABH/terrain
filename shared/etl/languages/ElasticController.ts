/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.
// tslint:disable:no-var-requires import-spacing strict-boolean-expressions

import * as Immutable from 'immutable';
const { List, Map } = Immutable;
import * as _ from 'lodash';

import { LanguageInterface } from 'shared/etl/languages/LanguageControllers';
import { ElasticTypes } from 'shared/etl/types/ETLElasticTypes';
import { FieldTypes, Languages } from 'shared/etl/types/ETLTypes';
import TypeUtil from 'shared/etl/TypeUtil';
import { TransformationEngine } from 'shared/transformations/TransformationEngine';
import EngineUtil from 'shared/transformations/util/EngineUtil';
import { KeyPath } from 'shared/util/KeyPath';
import * as yadeep from 'shared/util/yadeep';

import TransformationNodeType, { NodeOptionsType } from 'shared/transformations/TransformationNodeType';

class ElasticController implements LanguageInterface
{
  public language = Languages.Elastic;

  public changeFieldTypeSideEffects(engine: TransformationEngine, fieldId: number, newType)
  {
    const elasticProps = engine.getFieldProp(fieldId, List(['elastic']));
    if (elasticProps !== undefined)
    {
      const newProps = _.extend({}, elasticProps, {
        elasticType: ElasticTypes.Auto,
      });
      engine.setFieldProp(fieldId, List(['elastic']), newProps);
    }
  }

  public autodetectFieldTypes(engine: TransformationEngine, documents: List<object>)
  {
    // const docs = EngineUtil.preprocessDocuments(documents);
    // engine.getAllFieldIDs().forEach((id) =>
    // {
    //   if (engine.getFieldProp(id, List(['elastic', 'isPrimaryKey'])))
    //   {
    //     return;
    //   }
    //   const ikp = engine.getInputKeyPath(id);
    //   const okp = engine.getOutputKeyPath(id);

    //   let values = [];
    //   docs.forEach((doc) =>
    //   {
    //     const vals = yadeep.get(engine.transform(doc), okp);
    //     values = values.concat(vals);
    //   });
    //   const repType = EngineUtil.getRepresentedType(id, engine);
    //   if (repType === 'string')
    //   {
    //     const type = TypeUtil.getCommonElasticType(values);
    //     if (type === ElasticTypes.GeoPoint)
    //     {
    //       engine.appendTransformation(TransformationNodeType.CastNode, List([ikp]), { toTypename: 'object' });
    //       engine.setFieldType(id, 'object');
    //       const latField = engine.addField(ikp.push('lat'), 'number');
    //       const longField = engine.addField(ikp.push('lon'), 'number');
    //       engine.setOutputKeyPath(latField, okp.push('lat'));
    //       engine.setOutputKeyPath(longField, okp.push('lon'));
    //       EngineUtil.castField(engine, latField, 'number');
    //       EngineUtil.castField(engine, longField, 'number');
    //     }
    //     engine.setFieldProp(id, List(['elastic', 'elasticType']), type);
    //   }
    //   else if (repType === 'number')
    //   {
    //     const type = TypeUtil.getCommonElasticNumberType(values);
    //     engine.setFieldProp(id, List(['elastic', 'elasticType']), type);
    //   }
    // });
  }
}

export default new ElasticController();
