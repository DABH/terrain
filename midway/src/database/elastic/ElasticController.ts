/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2018 Terrain Data, Inc.

import * as Tasty from '../../tasty/Tasty';

import DatabaseController from '../DatabaseController';
import ElasticClient from './client/ElasticClient';
import ElasticConfig from './ElasticConfig';
import ElasticQueryHandler from './query/ElasticQueryHandler';
import ElasticDB from './tasty/ElasticDB';

/**
 * The central controller for communicating with ElasticSearch.
 */
class ElasticController extends DatabaseController
{
  private client: ElasticClient;
  private tasty: Tasty.Tasty;
  private queryHandler: ElasticQueryHandler;
  private analyticsIndex: string;
  private analyticsType: string;

  constructor(config: ElasticConfig, id: number, name: string, analyticsIndex?: string, analyticsType?: string,
    Client: { new(controller: ElasticController, config: ElasticConfig): ElasticClient } = ElasticClient)
  {
    super('ElasticController', id, name);

    this.client = new Client(this, config);

    this.tasty = new Tasty.Tasty(
      this,
      new ElasticDB(this.client));

    this.queryHandler = new ElasticQueryHandler(this);

    if (analyticsIndex !== undefined)
    {
      this.analyticsIndex = analyticsIndex;
    }

    if (analyticsType !== undefined)
    {
      this.analyticsType = analyticsType;
    }
  }

  public getClient(): ElasticClient
  {
    return this.client;
  }

  public getTasty(): Tasty.Tasty
  {
    return this.tasty;
  }

  public getQueryHandler(): ElasticQueryHandler
  {
    return this.queryHandler;
  }

  public getAnalyticsDB(): object
  {
    return {
      index: this.analyticsIndex,
      type: this.analyticsType,
    };
  }
}

export default ElasticController;
