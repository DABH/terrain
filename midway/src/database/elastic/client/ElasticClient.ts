/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import * as Elastic from 'elasticsearch';

import ElasticConfig from '../ElasticConfig';
import ElasticController from '../ElasticController';
import ElasticCluster from './ElasticCluster';
import ElasticIndices from './ElasticIndices';

/**
 * An client which acts as a selective isomorphic wrapper around
 * the elastic.js API.
 */
class ElasticClient
{
  public cluster: ElasticCluster;
  public indices: ElasticIndices;

  private controller: ElasticController;
  private config: ElasticConfig;
  private delegate: Elastic.Client;

  constructor(controller: ElasticController, config: ElasticConfig)
  {
    this.controller = controller;

    // Do not reuse objects to configure the elasticsearch Client class:
    // https://github.com/elasticsearch/elasticsearch-js/issues/33
    this.config = JSON.parse(JSON.stringify(config));
    this.delegate = new Elastic.Client(this.config);

    this.cluster = new ElasticCluster(controller, this.delegate);
    this.indices = new ElasticIndices(controller, this.delegate);
  }

  /**
   * https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#api-bulk
   */
  public bulk(params: Elastic.BulkIndexDocumentsParams, callback: (error: any, response: any) => void): void
  {
    // this.log('bulk', params);
    this.delegate.bulk(params, callback);
  }

  /**
   * https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#api-delete
   */
  public delete(params: Elastic.DeleteDocumentParams,
    callback: (error: any, response: Elastic.DeleteDocumentResponse) => void): void
  {
    this.log('delete', params);
    this.delegate.delete(params, callback);
  }

  /**
   * https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#api-deletetemplate
   */
  public deleteTemplate(params: Elastic.DeleteTemplateParams, callback: (error: any, response: any) => void): void
  {
    this.log('deleteTemplate', params);
    this.delegate.deleteTemplate(params, callback);
  }

  /**
   * https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#api-gettemplate
   */
  public getTemplate(params: Elastic.GetTemplateParams, callback: (error: any, response: any) => void): void
  {
    this.log('getTemplate', params);
    this.delegate.getTemplate(params, callback);
  }

  /**
   * https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#api-index
   */
  public index<T>(params: Elastic.IndexDocumentParams<T>, callback: (error: any, response: any) => void): void
  {
    this.log('index', params);
    this.delegate.index(params, callback);
  }

  /**
   * https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#api-putscript
   */
  public putScript(params: Elastic.PutScriptParams, callback: (err: any, response: any, status: any) => void): void
  {
    this.log('putScript', params);
    this.delegate.putScript(params, callback);
  }

  /**
   * https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#api-puttemplate
   */
  public putTemplate(params: Elastic.PutTemplateParams, callback: (err: any, response: any, status: any) => void): void
  {
    this.log('putTemplate', params);
    this.delegate.putTemplate(params, callback);
  }

  /**
   * https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#api-search
   */
  public search<T>(params: Elastic.SearchParams,
    callback: (error: any, response: Elastic.SearchResponse<T>) => void): void
  {
    this.log('search', params);
    this.delegate.search(params, callback);
  }

  public getDelegate(): Elastic.Client
  {
    return this.delegate;
  }

  private log(methodName: string, info: any)
  {
    this.controller.log('ElasticClient.' + methodName, info);
  }

}

export default ElasticClient;
