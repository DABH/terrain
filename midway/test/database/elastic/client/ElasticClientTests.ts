/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2018 Terrain Data, Inc.

import * as winston from 'winston';

import SharedUtil from '../../../../../shared/Util';
import * as Utils from '../../TestUtil';

import ElasticClient from '../../../../src/database/elastic/client/ElasticClient';
import ElasticConfig from '../../../../src/database/elastic/ElasticConfig';
import ElasticController from '../../../../src/database/elastic/ElasticController';
import PrefixedElasticController from '../../../../src/database/elastic/PrefixedElasticController';

import { MSearchResponse, SearchResponse } from 'elasticsearch';

let elasticController: ElasticController;
let elasticClient: ElasticClient;

function getExpectedFile(): string
{
  return __filename.split('.')[0] + '.expected';
}

beforeAll(async (done) =>
{
  // TODO: get rid of this monstrosity once @types/winston is updated.
  (winston as any).level = 'debug';
  const config: ElasticConfig = {
    hosts: ['http://localhost:9200'],
  };

  await new Promise((resolve, reject) =>
  {
    new ElasticController(config, 0, 'ElasticClientTests2').getClient().bulk(
      {
        body: [
          { index: { _index: 'bcd.secret', _type: 'data', _id: '13333337' } },
          { msg: 'This is a secret' },
          { index: { _index: 'bcd.secret', _type: 'data', _id: '133333372' } },
          { msg: 'This is a secret2' },
        ],
      },
      SharedUtil.promise.makeCallback(resolve, reject),
    );
  });

  elasticController = new PrefixedElasticController(config, 0, 'ElasticClientTests', undefined, undefined, 'abc.');
  elasticClient = elasticController.getClient();
  done();
});

test('elastic health', async (done) =>
{
  const result = await new Promise((resolve, reject) =>
  {
    elasticClient.cluster.health(
      {},
      SharedUtil.promise.makeCallback(resolve, reject));
  });
  winston.info(JSON.stringify(result));
  done();
});

test('search', async (done) =>
{
  try
  {
    const result: any = await new Promise((resolve, reject) =>
    {
      elasticClient.search(
        {
          index: 'movies',
          type: 'data',
          body: {
            sort: [{ revenue: 'desc' }, { movieid: 'asc' }],
          },
          size: 1,
        },
        SharedUtil.promise.makeCallback(resolve, reject));
    });
    winston.info(JSON.stringify(result, null, 2));
    await Utils.checkResults(getExpectedFile(), 'search', result.hits);
  }
  catch (e)
  {
    fail(e);
  }

  done();
});

test('indices.getMapping', async (done) =>
{
  try
  {
    const result = await new Promise((resolve, reject) =>
    {
      elasticClient.indices.getMapping(
        {},
        SharedUtil.promise.makeCallback(resolve, reject));
    });
    winston.info(JSON.stringify(result, null, 2));
    await Utils.checkResults(getExpectedFile(), 'indices.getMapping', result);
  }
  catch (e)
  {
    fail(e);
  }

  done();
});

test('putScript', async (done) =>
{
  try
  {
    await new Promise((resolve, reject) =>
    {
      elasticClient.putScript(
        {
          id: 'terrain_test_movie_profit',
          lang: 'painless',
          body: `return doc['revenue'].value - doc['budget'].value;`,
        },
        SharedUtil.promise.makeCallback(resolve, reject));
    });

    const result: any = await new Promise((resolve, reject) =>
    {
      elasticClient.search(
        {
          index: 'movies',
          type: 'data',
          body: {
            sort: {
              _script: {
                type: 'number',
                order: 'desc',
                script: {
                  id: 'terrain_test_movie_profit',
                  params: {},
                },
              },
            },
          },
          size: 1,
        },
        SharedUtil.promise.makeCallback(resolve, reject));
    });
    winston.info(JSON.stringify(result, null, 2));
    await Utils.checkResults(getExpectedFile(), 'putScript', result.hits);
  }
  catch (e)
  {
    fail(e);
  }

  done();
});

test('prefix isolation', async (done) =>
{
  const checkHit = (hit) =>
  {
    expect(hit._index).not.toMatch('secret');
    expect(hit._id).not.toMatch('13333337');
    expect(hit._source.msg).not.toMatch('secret');
  };

  const expectNoSecret = async (method, params) =>
  {
    (await new Promise<SearchResponse<any>>((resolve, reject) =>
    {
      elasticClient[method](
        params,
        SharedUtil.promise.makeCallback(resolve, reject),
      );
    })).hits.hits.forEach(checkHit);
  };

  const expectNoSecretMsearch = async (params) =>
  {
    (await new Promise<MSearchResponse<any>>((resolve, reject) =>
    {
      elasticClient.msearch(
        params,
        SharedUtil.promise.makeCallback(resolve, reject),
      );
    })).responses.forEach((res) => res.hits.hits.forEach(checkHit));
  };

  await expectNoSecret('search',
    {
      body: {
        query: {
          term: {
            _id: '13333337',
          },
        },
      },
    },
  );

  await expect(expectNoSecret('search',
    {
      index: 'bcd.secret',
      body: {
        query: {
          term: {
            _id: '13333337',
          },
        },
      },
    },
  )).rejects.toThrow(/index_not_found_exception/g);

  const scrollId = (await new Promise<SearchResponse<any>>((resolve, reject) =>
  {
    elasticClient.search(
      {
        scroll: '1m',
        body: {
          query: {
            terms: {
              _id: [ '13333337', '133333372' ],
            },
          },
        },
        size: 1,
      },
      SharedUtil.promise.makeCallback(resolve, reject));
  }))._scroll_id;
  await expectNoSecret('scroll',
    {
      scroll_id: scrollId,
    },
  );

  await expectNoSecretMsearch(
    {
      body: [
        {},
        {
          query: {
            term: {
              _id: '13333337',
            },
          },
        },
      ],
    },
  );

  done();
});
