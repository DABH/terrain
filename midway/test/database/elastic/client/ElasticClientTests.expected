{
  "search" : {
    "max_score" : null,
    "hits" : [
      {
        "_index" : "movies",
        "_type" : "data",
        "_id" : "72998",
        "_score" : null,
        "_source" : {
          "overview" : "In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.",
          "votecount" : 8218,
          "posterpath" : "/tcqb9NHdw9SWs2a88KCDD4V8sVR.jpg",
          "runtime" : 162,
          "movieid" : 72998,
          "language" : "en",
          "releasedate" : "2009-12-10T00:00:00.000Z",
          "voteaverage" : 7.099999904632568,
          "title" : "Avatar (2009)",
          "revenue" : 2100000000,
          "backdroppath" : "/5XPPB44RQGfkBrbJxmtdndKz05n.jpg",
          "genres" : "Action|Adventure|Sci-Fi|IMAX",
          "popularity" : 8.273819923400879,
          "@version" : "1",
          "tagline" : "Enter the World of Pandora.",
          "status" : "Released",
          "budget" : 237000000,
          "homepage" : "http://www.avatarmovie.com/"
        },
        "sort" : [
          2100000000,
          72998
        ]
      }
    ]
  },
  "indices.getMapping" : {
    "movies" : {
      "mappings" : {
        "data" : {
          "properties" : {
            "@version" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "backdroppath" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "budget" : {
              "type" : "long"
            },
            "genres" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "homepage" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "language" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "movieid" : {
              "type" : "long"
            },
            "overview" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "popularity" : {
              "type" : "float"
            },
            "posterpath" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "releasedate" : {
              "type" : "date"
            },
            "revenue" : {
              "type" : "long"
            },
            "runtime" : {
              "type" : "long"
            },
            "status" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "tagline" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "title" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "voteaverage" : {
              "type" : "float"
            },
            "votecount" : {
              "type" : "long"
            }
          }
        }
      }
    }
  },
  "putScript" : {
    "max_score" : null,
    "hits" : [
      {
        "_index" : "movies",
        "_type" : "data",
        "_id" : "72998",
        "_score" : null,
        "_source" : {
          "overview" : "In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.",
          "votecount" : 8218,
          "posterpath" : "/tcqb9NHdw9SWs2a88KCDD4V8sVR.jpg",
          "runtime" : 162,
          "movieid" : 72998,
          "language" : "en",
          "releasedate" : "2009-12-10T00:00:00.000Z",
          "voteaverage" : 7.099999904632568,
          "title" : "Avatar (2009)",
          "revenue" : 2100000000,
          "backdroppath" : "/5XPPB44RQGfkBrbJxmtdndKz05n.jpg",
          "genres" : "Action|Adventure|Sci-Fi|IMAX",
          "popularity" : 8.273819923400879,
          "@version" : "1",
          "tagline" : "Enter the World of Pandora.",
          "status" : "Released",
          "budget" : 237000000,
          "homepage" : "http://www.avatarmovie.com/"
        },
        "sort" : [
          1863000000
        ]
      }
    ]
  }
}
